Lib/linecache.py

```diff
@@ -135,7 +135,7 @@ def updatecache(filename, module_globals=None):
     try:
         with tokenize.open(fullname) as fp:
             lines = fp.readlines()
-    except OSError:
+    except (OSError, UnicodeDecodeError, SyntaxError):
```

Comment:
This effectively means repeated traceback line lookup code is silently slower in any of these error cases. UnicodeDecodeError and SyntaxError being ones that involve actual IO and parsing logic of some complexity. Regardless, this does fix the issue at hand.

While I wouldn't add this idea in a bugfix backport: caching the negative here would be valuable. We don't expect source files to come and go during a processes lifetime. If it isn't valid one time when needed, we may as well assume that it never is. That'll require deciding how to store the negative in the cache and modifying getlines() to understand that.

Reply:
I agree, the cache becoming stale relative to the disk is a problem we have already with found files, people occasionally create an issue about that.

So we backport this and then do that for 3.12.

Reply:
+1